coll/jq_test.go

```diff
@@ -0,0 +1,152 @@
+					"category": "fiction",
+					"author":   "Herman Melville",
+					"title":    "Moby Dick",
+					"isbn":     "0-553-21311-3",
+					"price":    8.99,
+				},
+				map[string]interface{}{
+					"category": "fiction",
+					"author":   "J. R. R. Tolkien",
+					"title":    "The Lord of the Rings",
+					"isbn":     "0-395-19395-8",
+					"price":    22.99,
+				},
+			},
+			"bicycle": map[string]interface{}{
+				"color": "red",
+				"price": 19.95,
+			},
+		},
+	}
+	out, err := JQ(".store.bicycle.color", in)
+	assert.NoError(t, err)
+	assert.Equal(t, "red", out)
+
+	out, err = JQ(".store.bicycle.price", in)
+	assert.NoError(t, err)
+	assert.Equal(t, 19.95, out)
+
+	out, err = JQ(".store.bogus", in)
+	assert.NoError(t, err)
+	assert.Nil(t, out)
+
+	_, err = JQ("{.store.unclosed", in)
+	assert.Error(t, err)
+
+	out, err = JQ(".store", in)
+	assert.NoError(t, err)
+	assert.EqualValues(t, in["store"], out)
+
+	out, err = JQ(".store.book[].author", in)
+	assert.NoError(t, err)
+	assert.Len(t, out, 4)
+	assert.Contains(t, out, "Nigel Rees")
+	assert.Contains(t, out, "Evelyn Waugh")
+	assert.Contains(t, out, "Herman Melville")
+	assert.Contains(t, out, "J. R. R. Tolkien")
+
+	out, err = JQ(".store.book[]|select(.price < 10.0 )", in)
+	assert.NoError(t, err)
+	expected := []interface{}{
+		map[string]interface{}{
+			"category": "reference",
+			"author":   "Nigel Rees",
+			"title":    "Sayings of the Century",
+			"price":    8.95,
+		},
+		map[string]interface{}{
+			"category": "fiction",
+			"author":   "Herman Melville",
+			"title":    "Moby Dick",
+			"isbn":     "0-553-21311-3",
+			"price":    8.99,
+		},
+	}
+	assert.EqualValues(t, expected, out)
+
+	in = map[string]interface{}{
+		"a": map[string]interface{}{
+			"aa": map[string]interface{}{
+				"foo": map[string]interface{}{
+					"aaa": map[string]interface{}{
+						"aaaa": map[string]interface{}{
+							"bar": 1234,
+						},
+					},
+				},
+			},
+			"ab": map[string]interface{}{
+				"aba": map[string]interface{}{
+					"foo": map[string]interface{}{
+						"abaa": true,
+						"abab": "baz",
+					},
+				},
+			},
+		},
+	}
+	out, err = JQ(`tostream|select((.[0]|index("foo")) and (.[0][-1]!="foo") and (.[1])) as $s|($s[0]|index("foo")+1) as $ind|($ind|truncate_stream($s)) as $newstream|$newstream|reduce . as [$p,$v] ({};setpath($p;$v))|add`, in)
+	assert.NoError(t, err)
+	assert.Len(t, out, 3)
+	assert.Contains(t, out, map[string]interface{}{"aaaa": map[string]interface{}{"bar": 1234}})
+	assert.Contains(t, out, true)
+	assert.Contains(t, out, "baz")
+
+	type bicycleType struct {
+		Color string
+	}
+	type storeType struct {
+		Bicycle *bicycleType
+		safe    interface{}
+	}
+
+	structIn := &storeType{
+		Bicycle: &bicycleType{
+			Color: "red",
+		},
+		safe: "hidden",
+	}
+
+	// TODO: Check if this is a valid test case (taken from jsonpath_test.go) since the struct
+	// had to be converted to JSON and parsed from it again to be able to process using gojq.
```

Comment:
🤔

I suppose this is because JQ can't be used with arbitrary structs or other types, which is perhaps something we should allow...

In the same vein, it might be useful to be able to pass the template context ([`gomplate.tmplctx`](https://github.com/hairyhenderson/gomplate/blob/main/context.go#L12)) as the input for something like:

```console
$ gomplate -c books=https://openlibrary.org/subjects/fantasy.json -i '{{ jq `.works[].title` . }}'
```

Right now this fails with:

```
executing JQ failed: %!v(PANIC=Error method: invalid type: *gomplate.tmplctx (&map[books:map[key:/subjects/fantasy...
```

What's interesting in that particular case is that `gomplate.tmplctx` is a `map[string]interface{}`, which is in theory OK.

I'm going to hack on this a bit and make a commit if I can figure It out...

Reply:
Ok, I've gotten this to work:

```console
$ bin/gomplate -c books=https://openlibrary.org/subjects/fantasy.json \
-i '{{ range jq `.books.works[].title` . -}}
title: {{.}}
{{end}}'
title: Alice's Adventures in Wonderland
title: Gulliver's Travels
title: Treasure Island
title: Through the Looking-Glass
title: A Midsummer Night's Dream
title: Il principe
title: The Wonderful Wizard of Oz
title: Avventure di Pinocchio
title: Alice's Adventures in Wonderland / Through the Looking Glass
title: Five Children and It
title: The Hobbit
title: Harry Potter and the Philosopher's Stone
```

Note that I've referenced the `books` datasource in the `.` context passed through as an argument.